fakeserver/fakeserver_integration_test.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"testing"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/zap"
+)
+
+var token = createSignedToken()
+
+func createSignedToken() string {
+	aud := "192.168.49.1:50051"
+	token := "token1"
+	// Example of generating a JWT with an "aud" claim
+	jwtToken := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
+		"sub": token,
+		"aud": []string{aud},
+		"exp": time.Now().Add(time.Hour * 72).Unix(),
+	})
+	// Just using "secret" for test signing
+	mySigningKey := []byte("secret")
+
+	// Sign and get the complete encoded token as a string
+	// nosemgrep: jwt.hardcoded-jwt-key
+	signedToken, err := jwtToken.SignedString(mySigningKey)
+	if err != nil {
+		logger.Error("Token could not be signed with fake secret key")
+	}
+	return signedToken
+
+}
+
+// Simulating a client interacting with the gRPC server
+func TestFakeServerConnectionSuccesfulAndRetry(t *testing.T) {
+	logger := zap.NewNop()
+	// Setup: Start the FakeServer
+	fakeServer := &FakeServer{
+		address:     "0.0.0.0:50051",
+		httpAddress: "0.0.0.0:50053",
+		token:       token,
+		logger:      logger, // Use a no-op logger for testing
+		state:       &ServerState{ConnectionSuccessful: false},
+	}
+
+	// Start the server
+	err := fakeServer.start()
+	assert.NoError(t, err, "Failed to start the FakeServer")
+
+	// Cleanup: Stop the server after the test
+	defer fakeServer.stop()
+
+	// Allow the server a bit of time to start
+	time.Sleep(2 * time.Second)
```

Comment:
Why this sleep? This seems brittle. Can't you use a WaitGroup or something like that?